### ISSUE TITLE
Unable to bulk insert entities with collection of owned entities

### ISSUE THREAD

**Issue by fdipuma** [borisdj/EFCore.BulkExtensions#695]
It seems I'm unable to do BulkInserts / BulkSaveChanges with EFCore.BulkExtensions 6.2.2 (Sql Server provider) and collections of Owned entities.

This is caused by multiple factors:
1. I receive a `NullReferenceException` when using `BulkSaveChanges`:
```
at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InnerGetDataTable[T](DbContext context, Type& type, IList`1 entities, TableInfo tableInfo) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 541
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.GetDataTable[T](DbContext context, Type type, IList`1 entities, SqlBulkCopy sqlBulkCopy, TableInfo tableInfo) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 418
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync)
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 104
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 31
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 209
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 294
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 119
   at EFCore.BulkExtensions.SqlBulkOperation.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SqlBulkOperation.cs:line 47
   at EFCore.BulkExtensions.DbContextBulkTransaction.ExecuteAsync[T](DbContext context, Type type, IList`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransaction.cs:line 90
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.InvokeBulkMethod(DbContext context, List`1 entities, Type entityType, String methodName, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 177
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 129
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 165
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 24
   at EFCore.BulkExtensions.DbContextBulkTransaction.ExecuteAsync[T](DbContext context, Type type, IList`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransaction.cs:line 66
```

This seems to be caused by this piece of code:
https://github.com/borisdj/EFCore.BulkExtensions/blob/5e15448cbe7b5dff5a90c87b77e870e75852c702/EFCore.BulkExtensions/SQLAdapters/SQLServer/SqlServerAdapter.cs#L529-L534

`ownedEntityType` is always null because `property.PropertyType` is a collection of owned entities (e.g. `List<Owned>`), but lookup is done like it was a single owned entity. This causes a `NullReferenceException` on the following line `ownedEntityType.GetProperties()`.

Since collections of owned entities are stored in a separate table it should be handled in the same way.

2. Tried adding this line to *skip* owned entities when not found inside the model:
```csharp
                    var ownedEntityType = context.Model.FindEntityType(property.PropertyType);
                    if (ownedEntityType == null)
                    {
                        ownedEntityType = context.Model.GetEntityTypes().SingleOrDefault(x => x.ClrType == property.PropertyType && x.Name.StartsWith(entityType.Name + "." + property.Name + "#"));
                    }
                    if (ownedEntityType == null)
                    {
                        continue; // workaround
                    }
                    var ownedEntityProperties = ownedEntityType.GetProperties().ToList();
```
But this lead to another issue:

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'index')
   at System.Linq.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument)
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.GetDiscriminatorColumn(TableInfo tableInfo) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 731
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InnerGetDataTable[T](DbContext context, Type& type, IList`1 entities, TableInfo tableInfo) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 462
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.GetDataTable[T](DbContext context, Type type, IList`1 entities, SqlBulkCopy sqlBulkCopy, TableInfo tableInfo) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 418
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync)
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 104
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.InsertAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 31
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 209
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 294
   at EFCore.BulkExtensions.SQLAdapters.SQLServer.SqlOperationsServerAdapter.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SQLAdapters\SQLServer\SqlServerAdapter.cs:line 119
   at EFCore.BulkExtensions.SqlBulkOperation.MergeAsync[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\SqlBulkOperation.cs:line 47
   at EFCore.BulkExtensions.DbContextBulkTransaction.ExecuteAsync[T](DbContext context, Type type, IList`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransaction.cs:line 90
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.InvokeBulkMethod(DbContext context, List`1 entities, Type entityType, String methodName, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 177
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 129
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken, Boolean isAsync) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 165
   at EFCore.BulkExtensions.DbContextBulkTransactionSaveChanges.SaveChangesAsync(DbContext context, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransactionSaveChanges.cs:line 24
   at EFCore.BulkExtensions.DbContextBulkTransaction.ExecuteAsync[T](DbContext context, Type type, IList`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress, CancellationToken cancellationToken) in D:\Dev\Github\EFCore.BulkExtensions\EFCore.BulkExtensions\DbContextBulkTransaction.cs:line 66
   at Program.<Main>$(String[] args) in D:\Dev\Github\EFCore.BulkExtensions\OwnedCollectionsTest\Program.cs:line 24
   at Program.<Main>(String[] args)
```

Seems EFCore.BulkExtensions is trying to get a discriminator (but there is no discriminator on the Owned entity).

3. If treated like a common entity EFCore.BulkExtensions tries build the query with `context.Set<T>()` in this line:

https://github.com/borisdj/EFCore.BulkExtensions/blob/5e15448cbe7b5dff5a90c87b77e870e75852c702/EFCore.BulkExtensions/TableInfo.cs#L895-L910

But owned entities cannot be queryed using `context.Set<T>` directly.

So it seems to me that collection owned entities is not currently supported?

### Full reproducible example:

```csharp
using EFCore.BulkExtensions;
using Microsoft.EntityFrameworkCore;

const string DefaultConnectionString = "Server=(localdb)\\MSSQLLocalDB;Integrated Security=true;Initial Catalog=EfBulkTest;Trust Server Certificate=true;Encrypt=No;";

var optionsBuilder = new DbContextOptionsBuilder<MyDbContext>()
    .UseSqlServer(DefaultConnectionString);

using var context = new MyDbContext(optionsBuilder.Options);

await context.Database.EnsureCreatedAsync();

context.Set<Root>().AddRange(Enumerable.Range(0, 100).Select(i => new Root
{
    OwnedCollection = new List<Owned>
    {
        new()
        {
            OwnedField = i.ToString()
        }
    }
}));

await context.BulkSaveChangesAsync();

public class Root
{
    public int Id { get; set; }
    public List<Owned> OwnedCollection { get; set; }
}

public class Owned
{
    public string OwnedField { get; set; }
}
    
public class MyDbContext : DbContext
{
    public MyDbContext(DbContextOptions<MyDbContext> options) : base(options)
    {
        
    }
    
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Owned<Owned>();
        modelBuilder.Entity<Root>(o =>
        {
            o.HasKey(e => e.Id);
            o.OwnsMany(x => x.OwnedCollection, opt =>
            {
                opt.Property("Id").ValueGeneratedOnAdd();
            });
        });
    }
}
```

**Comment by borisdj** [comment]
For the moment you could try manual config as explained here: https://github.com/borisdj/EFCore.BulkExtensions/issues/114#issuecomment-803462928

**Comment by fdipuma** [comment]
Not sure I understand your comment, it seems to me that:
- The workaround proposed in #114 is only for individual Owned types (not collections)
- The second BulkInsert will fail, in this case, because the shadow property used as Id to the Owner table will not be set by the first BulkInsert
- `SqlOperationsServerAdapter.InsertAsync` will be executed in the same way and fail as stated on my third point in the OP

Thanks!

**Comment by borisdj** [comment]
I see now, then collection is not supported.

**Comment by LukeKolodziej** [comment]
I have also run into this issue.

**Comment by borisdj** [comment]
Should be fixed with v6.5.5.

**Comment by aupisarev** [comment]
BulkInsert in this same scenario doesn't work
